sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala

```diff
@@ -116,7 +116,9 @@ class RocksDBSuite extends SparkFunSuite {
     withDB(remoteDir, conf = conf) { db =>
       // Generate versions without cleaning up
       for (version <- 1 to 50) {
-        db.put(version.toString, version.toString)  // update "1" -> "1", "2" -> "2", ...
```

Comment:
This was something we could indicate the bug earlier... 

The test was incorrect. The test expects that compaction will happen, but if we don't "overwrite" the key and just put the new keys here, compaction does not happen, at least with Spark's default config. The reason compaction had happened is, writebatch wasn't cleaned up during loop, hence writes are accumulated into the same writebatch and it led to overwrites.

Reply:
how do we know this will definitely cause compaction?

Reply:
I'm not an expert of RocksDB, but looks like we could expect compaction according to the high level architecture.

https://github.com/facebook/rocksdb/wiki/RocksDB-Overview#3-high-level-architecture

During RocksDB.commit(), we write synchronously and flush synchronously, meaning that we will produce a new SST file for new writes. (As SST file is immutable, you can't append any new write against existing SST files.)

That said, for each loop, we write a bunch of overwritten keys, and produce a new SST file containing overwritten keys. L0 allows SST files to have overlapping key ranges, but it is no longer allowed in upper level, which is ensured via "compaction".

Reply:
I see, could you add this detail to the comment? it was really helpful for me to understand how we are ensuring compaction

Reply:
Sounds good, thanks for the suggestion!

Reply:
Or can we do this maybe
`db.putWithTtl(key.getBytes(), value.getBytes(), ttl)` and ensure the TTL has been exceeded by that time?

Reply:
No, the functionality is not here. We don't leverage TTL for the general state store operations.